sdk/core/azure-core/inc/http/stream.hpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+#include <stdio.h>
+
+namespace Azure { namespace Core { namespace Http {
+
+  // BodyStream is used to read data to/from a service
+  class BodyStream {
+  public:
+    static BodyStream* null;
+
+    // Returns the length of the data; used with the HTTP Content-Length header
+    virtual uint64_t Length() = 0;
+
+    // Resets the stream back to the beginning (for retries)
+    // Derived classes that send data in an HTTP request MUST override this and implement it
+    // properly.
+    virtual void Rewind()
+    {
+      throw "Not Implemented"; // TODO: Replace with best practice as defined by guideline
+    };
+
+    // Reads more data; throws if error/canceled
+    // return copied size
+    virtual uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count) = 0;
+
+    // Closes the stream; typically called after all data read or if an error occurs.
+    virtual void Close() = 0;
+
+    // Desstructor. Enables derived classes to call its destructor
+    virtual ~BodyStream() = 0;
+  };
+
+  class MemoryBodyStream : public BodyStream {
+  private:
+    uint8_t* m_ptr;
+    uint64_t offset;
+    uint64_t m_length;
+
+  public:
+    MemoryBodyStream(uint8_t* ptr, uint64_t length) : m_ptr(ptr), m_length(length)
+    {
+      this->offset = 0;
+    }
+
+    uint64_t Length() { return m_length; }
+
+    uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count)
+    {
+      uint64_t copy_length = std::min(count, (m_length - offset));
+      // Copy what's left or just the count
+      std::memcpy(buffer, m_ptr + offset, copy_length);
+      // move position
+      offset += copy_length;
+
+      return copy_length;
+    }
+
+    void Rewind() { offset = 0; }
+
+    void Close() {}
+  };
+
+  class FileBodyStream : public BodyStream {
+  private:
+    FILE* stream;
+    uint64_t length;
+
+  public:
+    FileBodyStream(FILE* stream)
+    {
+      // set internal fields
+      this->stream = stream;
+      // calculate size seeking end...
+      this->length = fseeko64(stream, 0, SEEK_END);
+      // seek back to beggin
+      this->Rewind();
+    }
+
+    // Rewind seek back to 0
+    void Rewind() { rewind(this->stream); }
+
+    uint64_t Read(/*Context& context, */ uint8_t* buffer, uint64_t count)
+    {
+      // do static cast here?
+      return (uint64_t)fread(buffer, 1, count, this->stream);
+    }
+
+    // close does nothing opp
+    void Close() { fclose(this->stream); }
```

Comment:
The `FILE*` is passed-in in constructor. That means you didn't open it but you close it.

I just feel that it's a little bit weird.

Reply:
It's just like in `MemoryBodyStream`, the buffer is passed in, you didn't allocate it but you free it in `Close()`.